ADR 239 - Avatar assembly definition (#242)
decentraland/adr: content/ADR-239-avatar-assembly-changes.md

---
adr: 239
date: 2023-06-01
title: Avatar Assembly Changes

> status: Review

Reviewer: This PR has a tide effect of potentially dozens/hundreds of hours across many teams.

Please update https://adr.decentraland.org/adr/ADR-57 (Avatar assembling instructions for visualization) to reflect the new changes in this same PR. That way we can use the diff in the previous one to update the:
- bevy implementation https://github.com/decentraland/bevy-explorer/pull/19/files
- babylon implementation https://github.com/decentraland/hammurabi/pull/31
- werable preview for marketplace & for third parties (DG mobile) https://wearable-preview.decentraland.org/?profile=default1
- docs site preview https://docs.decentraland.org/creator/emotes/emotes-overview/

Author: I thought old ADRs were supposed to be set in stone once approved, should I modify that ADR in this PR? Or how should I handle this? @menduz

Author: done

The decision-making capacity of creators is overlapping with the freedom of choice of players. When a player purchases a wearable, thinking it will combine well with what they already have in their inventory, they may be unpleasantly surprised that these are not compatible. Information about which other wearables are replaced or hidden is not clearly indicated in the marketplace or backpack.
There is no real weighty reason for a wearable to unequip another wearable, instead of hiding it to prevent clipping problems.

On the base of that the replacing functionality for new wearables will be removed, pre-existing wearables will see the replacing list being merged with the hiding list.

### Body fallback

Previously, by unequipping the upper_body, the lower_body, the feet or the eyebrows, caused the system to fallback to default wearables (t-shirt, pants and shoes).
This prevents, for example, players from going barefoot or without t-shirt.

The system will extent the concept of Body by removing the default fallback wearables.

> These pieces, which can be called "Bare feet", "Naked Top" and "Naked Bottom", are not considered wearables, but rather parts of the base body geometry that are exclusively displayed when no wearables are equipped and there are no other wearables hiding the relevant category.

Reviewer: A stronger algorithmic definition of these rules will be needed to implement the logic. What can we infer is only `name.isOneOf(["Bare feet", "Naked Top", "Naked Bottom"])` and that may not be enough

> The base body geometry and textures will change to make them more suitable for an environment for all audiences.

Reviewer: This comment has nothing to do with the ADR and can be omited
```suggestion
```

### Force Render

> To extend further players ability to use the wearables they have in all their glory. If an equipped wearable is hidden by the effect of another wearable, players can unhide them and override this instruction.

Reviewer: Clarity over narrative
```suggestion
To extend further players ability to use the wearables they have. If an equipped wearable is hidden by the effect of another wearable, players can unhide them and override this instruction.
```

This forced render effect is removed as soon as another wearable that re-hides that category is equipped, forcing the user to show it again.

In order to store the force render categories, it is proposed to add a new field to the player profile holding a list of overridden hide categories.
This list of string will hold the current overridden categories so that each player can render other players (and itself) according to the new rules.
As an example:

> ```

Reviewer: ````suggestion
```yaml
````

> "forceRender": ["mask","top_head","feet"]

Reviewer: Maybe, if we are still on-time, forcedCategories would be a better self-explanatory name
```suggestion
"forcedCategories": ["mask","top_head","feet"]
```

Author: sadly too late

Author: the functionality has already been added to unity renderer, browser interface and profile deploy validation.
Also I think that forceRender is descriptive enough, considering that something hidden is still equipped, so by adding it to the list we are forcing it to render even if hidden

```
This example means that when the avatar is going to be assembled the "mask", "top_head" and "feet" categories will be shown even if other wearables are hiding them.

Here's the avatar definition with the additional forceRender field that will be provided:

> ```

Reviewer: ````suggestion
```json
````

               "face256":"https://interconnected.online/content/contents/bafkreiddnf2ixknlhio4hk7ao47agiwkqs43cg3nkoyibckgmhz5iqn3he"
            },
            "eyes":{
               "color":{
                  "r":0.529411792755127,
                  "g":0.501960813999176,
                  "b":0.47058823704719543,
                  "a":1
               }
            },
            "hair":{
               "color":{
                  "r":0.10980392247438431,
                  "g":0.10980392247438431,
                  "b":0.10980392247438431,
                  "a":1
               }
            },
            "skin":{
               "color":{
                  "r":0.800000011920929,
                  "g":0.6078431606292725,
                  "b":0.46666666865348816,
                  "a":1
               }
            }
         },
         "ethAddress":"0x1hd7fl09yivcn7z6asc0a8af9c3b1ba28hdy65sd",
         "userId":"0x1hd7fl09yivcn7z6asc0a8af9c3b1ba28hdy65sd",
         "version":12,
         "hasConnectedWeb3":true
      }
   ]
}
```

### Avatar assembly suggested flow

With the points previously discovered, the suggested procedure to visualize an avatar is the following:

> * Request a user profile ( `https://catalyst/lambdas/profile/address` )

Reviewer: The lambdas endpoint is deprecated
```suggestion
* Request a user profile ( `POST https://content-server/content/active {"pointers": ["0xaddress"]}` )
```

> * Request the wearables listed in the profile wearables field, all wearables listed are the equipped ones, no matter their visibility. ( `https://catalyst/lambdas/collections/wearables?wearableId=urnId` )

Reviewer: The lambdas endpoint is deprecated

```suggestion
* Request the wearables listed in the profile wearables field, all wearables listed are the equipped ones, no matter their visibility. ( `POST https://content-server/content/active {"pointers": ["urn1", "urn2"]}` )
```
